dali/pipeline/executor/executor2/op_task.cc

```diff
@@ -0,0 +1,403 @@
+      auto tl = std::make_shared<TensorList<GPUBackend>>();
+      tl->set_order(ws.output_order());
+      if (device < 0)
+        CUDA_CALL(cudaGetDevice(&device));
+      tl->set_device_id(device);
+      ws.SetOutput(i, tl);
+    } else {
+      assert(!"Unreachable code - unknown backend.");
+    }
+  }
+
+  if (!skip_) {
+    DomainTimeRange tr("[DALI][OpTask] Setup " + GetOpDisplayName(node_->op->GetSpec()));
+    ApplyDefaultLayouts();
+    std::vector<OutputDesc> output_descs;
+    output_descs.reserve(nout);
+    // If Setup returns true, we must resize the outputs;
+    if (node_->op->Setup(output_descs, ws)) {
+      assert(output_descs.size() == static_cast<size_t>(nout));
+      for (int i = 0; i < nout; i++) {
+        if (ws.OutputIsType<CPUBackend>(i)) {
+          ws.Output<CPUBackend>(i).Resize(output_descs[i].shape, output_descs[i].type);
+        } else if (ws.OutputIsType<GPUBackend>(i)) {
+          auto &output = ws.Output<GPUBackend>(i);
+          output.Resize(output_descs[i].shape, output_descs[i].type);
+          CUDA_CALL(cudaMemsetAsync(output.raw_mutable_tensor(0), 128, output_descs[i].shape.num_elements() * TypeTable::GetTypeInfo(output_descs[i].type).size(), ws.stream()));
+        } else {
+          assert(!"Unreachable code - unknown backend.");
+        }
+      }
+    }
+  }
+}
+
+void OpTask::RunOp() {
+  if (!skip_) {
+    DomainTimeRange tr("[DALI][Executor] Run");
+    node_->op->Run(*ws_);
+    ResetInputLayouts();
+    PropagateSourceInfo(*ws_);
+  }
+  if (auto cpt = ws_->GetIterationData()->checkpoint) {
+    node_->op->SaveState(cpt->GetOpCheckpoint(node_->instance_name), ws_->output_order());
+  }
+  if (ws_->has_event() && ws_->has_stream())
+    CUDA_CALL(cudaEventRecord(ws_->event(), ws_->stream()));
+}
+
+void OpTask::SetWorkspaceInputs() {
+  int ti = 0;
+  assert(ws_->NumInput() + ws_->NumArgumentInput() == static_cast<int>(node_->inputs.size()));
+  auto order = ws_->output_order();
+  std::unordered_set<cudaEvent_t> events(ws_->NumInput() + ws_->NumArgumentInput());
+  for (int i = 0; i < ws_->NumInput(); i++, ti++) {
+    if (ws_->InputIsType<CPUBackend>(i)) {
+      auto inp = TaskInput<CPUBackend>(ti);
+      if (inp.event && inp.order != order)
+        events.insert(inp.event);
+      ws_->SetInput(i, inp.data);
+    } else {
+      assert(ws_->InputIsType<GPUBackend>(i));
+      auto inp = TaskInput<GPUBackend>(ti);
+      // If the output order of the operator is `host` then we don't wait for GPU
+      // inputs - they can't be accessed directly on host and the operator will
+      // have to issue some sort of synchronization if and when necessary.
+      // This optimization is essential to avoid oversynchronization
+      // when the operator needs to access the metadata only (e.g. getting the shape).
+      if (order.is_device() /*see comment above */ && inp.event && inp.order != order)
+        events.insert(inp.event);
+      ws_->SetInput(i, inp.data);
```

Comment:
Btw, did you approach the problem of data arriving with different order than expected (input order != `ws.output_order()`)?

Will this executor improve the situation compared to the previous one? What if we have consumers of such input that have different output order. Can we have some prologue/epilogue functionality on OpTask that handles such syncs for the user?

Reply:
The necessary synchronization is implemented in `SetWorkspaceInputs`. If we introduce yet another order (work_order vs output_order), we can simply change the target stream in that place.